### PR TITLE
PWX-27563: Setting stork as the default scheduler for px-csi-ext pods

### PR DESCRIPTION
**Pull request by pp511**
Signed-off-by: Priyanshu Pandey <ppandey@purestorage.com>

**What this PR does / why we need it**:
When CSI is enabled  PVC provisioning goes through px-csi-ext-* pods which is a deployment of 3 replicas. If these 3 PX nodes go offline none of the PVC creations would succeed even if the PX cluster is operational. In such cases we need to move the px-csi-ext pods to other nodes in the cluster.
This change will make Stork as the default scheduler for px-csi-ext-* pods. Change will be made in Stork to detect px-csi-ext pods are running on an offline PX node such that it can delete the csi-ext pod and schedule them on the online nodes.

**Which issue(s) this PR fixes** (optional)
Closes #https://portworx.atlassian.net/browse/PWX-27563

**Special notes for your reviewer**:
* Refer to https://portworx.atlassian.net/browse/PWX-27549 for details
* Tested  on dev setup that stork is the default scheduler for px-csi-ext-*  deployment.
* Ran the following job with these changes in this PR for regressions https://jenkins.pwx.dev.purestorage.com/job/test-op-csi-e2e/1/

### CODE REVIEW

**Comment by ggriffiths** [comment]
@pp511 do you know the behavior if stork is disabled in the storagecluster? Will k8s ignore the stork scheduler, or error out?

**Comment by pp511** [comment]
@ggriffiths If stork is disabled then then csi pods will not come online. Default k8 scheduler will not be used. 
We should use the default scheduler when stork is disabled. Right? Please confirm and I will update the PR.

**Comment by ggriffiths** [comment]
Yes, there might be cases where someone disables stork and enables CSI.

**Review by haibinxie — COMMENTED** [review]
If node goes offline, k8s default scheduler will reschedule the pod, doen't that work for us?

**Comment by pp511** [comment]
@ggriffiths Updated the PR to use Stork as the scheduler only when it is enabled in StorageCluster.
@haibinxie Currently Stork’s monitor controller monitors all the application pods that use Portworx PVCs and deletes such pods which are running on offline PX nodes. Something similar needs to be done for CSI pods where Stork needs to identify such px-csi-ext pods and use its scheduling logic even though there are no Portworx PVCs being used by them.

**Review by ggriffiths — APPROVED** [review]
LGTM

File: drivers/storage/portworx/component/csi.go [hunk 1]
```diff
@@ -666,6 +668,10 @@ func getCSIDeploymentSpec(
 		},
 	}
 
+	if pxutil.IsStorkEnabled(cluster) {
```
Comment (piyush-nimbalkar) [inline drivers/storage/portworx/component/csi.go:671]: This won't handle if stork is disabled at runtime. This is not a common scenario, but we should handle it.
Currently the CSI deployment gets recreated only if certain things change in the deployment based on the storagecluster spec. Check line https://github.com/libopenstorage/operator/pull/807/files#diff-41fa95b9f2df1987f95180171e28b9d77de07ce97cf8d5c34dea99ca3c54e0c2R506

We will have to add some custom logic there for the scheduler field or check with @haibinxie to use the custom DeepCopy function he introduced.

Comment (piyush-nimbalkar) [reply drivers/storage/portworx/component/csi.go:671]: Can you verify what happens if the stork scheduler is removed at runtime? Try deleting one of the stork pods and see what happens.

Comment (ggriffiths) [reply drivers/storage/portworx/component/csi.go:671]: Ah yes, we need to make sure the px-csi-ext is bounced when stork is disabled or enabled.

Comment (pp511) [reply drivers/storage/portworx/component/csi.go:669]: @piyush-nimbalkar Added the changes to handle the scenario when stork is enabled or disabled at run time.

File: pkg/util/util.go [hunk 2]
```diff
@@ -284,6 +287,22 @@ func HasNodeAffinityChanged(
 	return !reflect.DeepEqual(cluster.Spec.Placement.NodeAffinity, existingAffinity.NodeAffinity)
 }
 
+// HasSchedulerStateChanged checks if the stork has been enabled/disabled in the StorageCluster
+func HasSchedulerStateChanged(
+	cluster *corev1.StorageCluster,
+	previousSchedulerName string,
+) bool {
+	if cluster.Spec.Stork == nil {
+		return false
```
Comment (piyush-nimbalkar) [inline pkg/util/util.go:296]: If stork was enabled before and then the spec was set to empty, this should return true.

Comment (piyush-nimbalkar) [reply pkg/util/util.go:296]: Can you add UTs for this function?

Comment (pp511) [reply pkg/util/util.go]: Done

**Review by piyush-nimbalkar — APPROVED** [review]

**Comment by piyush-nimbalkar** [comment]
> If node goes offline, k8s default scheduler will reschedule the pod, doen't that work for us?

@haibinxie here we are talking about portworx node going offline. Kubernetes won't reschedule the csi ext pod in that case.

**Pull request merged**